agent/grpc-external/services/peerstream/stream_test.go

```diff
@@ -869,6 +868,179 @@ func TestStreamResources_Server_CARootUpdates(t *testing.T) {
+	require.Equal(t, pbpeerstream.TypeURLPeeringTrustBundle, receiveRoots.GetResponse().ResourceURL)
+
+	testutil.RunStep(t, "new stream gets tracked", func(t *testing.T) {
+		retry.Run(t, func(r *retry.R) {
+			status, ok := srv.StreamStatus(peerID)
+			require.True(r, ok)
+			require.True(r, status.Connected)
+		})
+	})
+
+	testutil.RunStep(t, "sends first heartbeat", func(t *testing.T) {
+		retry.RunWith(&retry.Timer{
+			Timeout: outgoingHeartbeatInterval * 2,
+			Wait:    outgoingHeartbeatInterval / 2,
+		}, t, func(r *retry.R) {
+			heartbeat, err := client.Recv()
+			require.NoError(t, err)
+			require.NotNil(t, heartbeat.GetHeartbeat())
+		})
+	})
+
+	testutil.RunStep(t, "sends second heartbeat", func(t *testing.T) {
+		retry.RunWith(&retry.Timer{
+			Timeout: outgoingHeartbeatInterval * 2,
+			Wait:    outgoingHeartbeatInterval / 2,
+		}, t, func(r *retry.R) {
+			heartbeat, err := client.Recv()
+			require.NoError(t, err)
+			require.NotNil(t, heartbeat.GetHeartbeat())
+		})
+	})
+}
+
+// Test that as long as the server receives heartbeats it keeps the connection open.
+func TestStreamResources_Server_KeepsConnectionOpenWithHeartbeat(t *testing.T) {
+	it := incrementalTime{
+		base: time.Date(2000, time.January, 1, 0, 0, 0, 0, time.UTC),
+	}
+	incomingHeartbeatTimeout := 10 * time.Millisecond
+
+	srv, store := newTestServer(t, func(c *Config) {
+		c.Tracker.SetClock(it.Now)
+		c.incomingHeartbeatTimeout = incomingHeartbeatTimeout
+	})
+
+	p := writePeeringToBeDialed(t, store, 1, "my-peer")
+	require.Empty(t, p.PeerID, "should be empty if being dialed")
+	peerID := p.ID
+
+	// Set the initial roots and CA configuration.
+	_, _ = writeInitialRootsAndCA(t, store)
+
+	client := makeClient(t, srv, peerID)
+
+	// TODO(peering): test fails if we don't drain the stream with this call because the
+	// server gets blocked sending the termination message. Figure out a way to let
+	// messages queue and filter replication messages.
+	receiveRoots, err := client.Recv()
+	require.NoError(t, err)
+	require.NotNil(t, receiveRoots.GetResponse())
+	require.Equal(t, pbpeerstream.TypeURLPeeringTrustBundle, receiveRoots.GetResponse().ResourceURL)
+
+	testutil.RunStep(t, "new stream gets tracked", func(t *testing.T) {
+		retry.Run(t, func(r *retry.R) {
+			status, ok := srv.StreamStatus(peerID)
+			require.True(r, ok)
+			require.True(r, status.Connected)
+		})
+	})
+
+	heartbeatMsg := &pbpeerstream.ReplicationMessage{
+		Payload: &pbpeerstream.ReplicationMessage_Heartbeat_{
+			Heartbeat: &pbpeerstream.ReplicationMessage_Heartbeat{}}}
+
+	// Set up a goroutine to send the heartbeat every 1/2 of the timeout.
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	go func() {
+		// This is just a do while loop. We want to send the heartbeat right away to start
+		// because the test setup above takes some time and we might be close to the heartbeat
+		// timeout already.
+		for {
+			require.NoError(t, client.Send(heartbeatMsg))
```

Comment:
Using the `t` within a go routine that isn't the test function executed by the go test framework make `go test` angry. It outputs warnings about this being bad and potentially could cause issues with things modify test state after the test has completed.

I think the general approach for getting around this is to create a `chan error` and have the go routine send to that chan any error it may encounter or nil. Then the main test func would recv from that chan (you would also need to manually call the cancel func earlier) and check the value for a non-nil error.

Reply:
Awesome catch thanks for that tip! I added the errCh and I think I'm handling it as suggested.